File: backend/app/game/team_metrics.py
"""Team aggregate scores and qualitative metrics from player stats."""

from __future__ import annotations

from typing import Sequence


def team_average_progress(progresses: Sequence[float]) -> float:
    if not progresses:
        return 0.0
    return sum(progresses) / len(progresses)


def teamwork_score(progresses: Sequence[float]) -> float:
    """Tighter progress spread within the team → higher teamwork."""
    if len(progresses) < 2:
        return 88.0
    mean_p = sum(progresses) / len(progresses)
    var = sum((x - mean_p) ** 2 for x in progresses) / len(progresses)
    stdev = var**0.5
    return max(0.0, min(100.0, 100.0 * (1.0 - min(1.0, stdev * 3.8))))


def consistency_score(accuracies: Sequence[float]) -> float:
    if not accuracies:
        return 0.0
    return max(0.0, min(100.0, 100.0 * (sum(accuracies) / len(accuracies))))


def communication_efficiency_score(
    keystrokes: Sequence[int],
    errors: Sequence[int],
    relay_passes: int,
    member_count: int,
) -> float:
    """Keystroke hit-rate scaled by relay handoffs (proxy for coordination)."""
    tk = sum(keystrokes)
    if tk <= 0:
        base = 0.75
    else:
        er = sum(errors)
        base = max(0.0, min(1.0, (tk - er) / tk))
    bonus = 1.0 + min(0.4, relay_passes * 0.06 / max(1, member_count))
    return max(0.0, min(100.0, 100.0 * base * bonus))

File: server/game/team-metrics.ts
export function teamAverageProgress(progresses: number[]): number {
  if (!progresses.length) return 0;
  return progresses.reduce((a, b) => a + b, 0) / progresses.length;
}

export function teamworkScore(progresses: number[]): number {
  if (progresses.length < 2) return 88;
  const mean = progresses.reduce((a, b) => a + b, 0) / progresses.length;
  const variance =
    progresses.reduce((s, x) => s + (x - mean) ** 2, 0) / progresses.length;
  const stdev = Math.sqrt(variance);
  return Math.max(0, Math.min(100, 100 * (1 - Math.min(1, stdev * 3.8))));
}

export function consistencyScore(accuracies: number[]): number {
  if (!accuracies.length) return 0;
  return Math.max(0, Math.min(100, (100 * accuracies.reduce((a, b) => a + b, 0)) / accuracies.length));
}

export function communicationEfficiencyScore(
  keystrokes: number[],
  errors: number[],
  relayPasses: number,
  memberCount: number
): number {
  const tk = keystrokes.reduce((a, b) => a + b, 0);
  let base: number;
  if (tk <= 0) base = 0.75;
  else {
    const er = errors.reduce((a, b) => a + b, 0);
    base = Math.max(0, Math.min(1, (tk - er) / tk));
  }
  const bonus = 1 + Math.min(0.4, (relayPasses * 0.06) / Math.max(1, memberCount));
  return Math.max(0, Math.min(100, 100 * base * bonus));
}
